src/main/java/org/opensearch/ad/feature/SearchFeatureDao.java

```diff
@@ -190,67 +233,211 @@ public void getHighestCountEntities(AnomalyDetector detector, long startTime, lo
+            AnomalyDetector detector,
+            SearchSourceBuilder searchSourceBuilder,
+            long expirationEpochMs
+        ) {
+            this.listener = listener;
+            this.detector = detector;
+            this.topEntities = new ArrayList<>();
+            this.searchSourceBuilder = searchSourceBuilder;
+            this.expirationEpochMs = expirationEpochMs;
+        }
+
+        @Override
+        public void onResponse(SearchResponse response) {
+            try {
+                Aggregations aggs = response.getAggregations();
+                if (aggs == null) {
+                    listener.onResponse(topEntities);
+                    return;
+                }
+
+                Aggregation aggrResult = aggs.get(AGG_NAME_TOP);
+                if (aggrResult == null) {
+                    listener.onFailure(new IllegalArgumentException("Fail to find valid aggregation result"));
+                    return;
+                }
+
+                if (detector.getCategoryField().size() == 1) {
+                    topEntities = ((Terms) aggrResult)
+                        .getBuckets()
+                        .stream()
+                        .map(bucket -> bucket.getKeyAsString())
+                        .collect(Collectors.toList())
+                        .stream()
+                        .map(
+                            entityValue -> Entity
+                                .createSingleAttributeEntity(detector.getDetectorId(), detector.getCategoryField().get(0), entityValue)
+                        )
+                        .collect(Collectors.toList());
+                    listener.onResponse(topEntities);
+                } else {
+                    CompositeAggregation compositeAgg = (CompositeAggregation) aggrResult;
+                    List<Entity> pageResults = compositeAgg
+                        .getBuckets()
+                        .stream()
+                        .filter(bucket -> bucket.getDocCount() >= minimumDocCountForPreview)
+                        .map(bucket -> Entity.createEntityByReordering(detector.getDetectorId(), bucket.getKey()))
+                        .collect(Collectors.toList());
+                    // we only need at most maxEntitiesForPreview
+                    int amountToWrite = maxEntitiesForPreview - topEntities.size();
+                    for (int i = 0; i < amountToWrite && i < pageResults.size(); i++) {
+                        topEntities.add(pageResults.get(i));
+                    }
+                    Map<String, Object> afterKey = compositeAgg.afterKey();
+                    if (topEntities.size() >= maxEntitiesForPreview || afterKey == null || expirationEpochMs < clock.millis()) {
+                        listener.onResponse(topEntities);
+                    } else {
+                        updateSourceAfterKey(afterKey, searchSourceBuilder);
+                        client
+                            .search(
```

Comment:
For my understanding, why is another search call being made here conditionally, but not in line 405?

Reply:
because at line 405, we are using terms aggregation and one call is enough.  Here we are using composite queries that are not a one-shot deal, like scrolling.